Storage/src/Connection/Rest.php

```diff
@@ -656,183 +658,69 @@ private function sendWithRetry($resource, $method, array $args)
+        return $this->traitSend($resource, $method, $args);
     }
 
     /**
-     * Adds the callback methods to $args which amends retry hash and attempt
-     * count to the headers.
+     * Adds the retry headers to $args which amends retry hash and attempt
+     * count to the required header.
      * @param array $args
-     *
      * @return array
      */
-    private function addRetryHeaderCallbacks(array $args)
+    private function addRetryHeaderLogic(array $args)
     {
         $requestHash = Uuid::uuid4()->toString();
-        $args['restOnRetryExceptionFunction'] = function (
-            \Exception $e,
-            $currentAttempt,
-            &$arguments
-        ) use ($requestHash) {
-            // Since we the the last attempt number here, so incrementing it
-            // to get the current attempt count.
-            // We're adding a '2' and not '1' as we need to incorporate the initial
-            // request as well.
-            $this->updateRetryHeaders(
-                $arguments,
-                $requestHash,
-                $currentAttempt + 2
-            );
-        };
+        $invocationIdHeaderValue = 'gccl-invocation-id/' . $requestHash;
+        $attempCountKey = 'gccl-attempt-count';
+        // Would be applied later in ExponentialBackoff>execute() after reading
+        // existing header value from Request object.
+        $args['restOptions']['retryHeaders'] = [
+            $invocationIdHeaderValue,
+            sprintf('%s/%d', $attempCountKey, 1)
+        ];
 
-        $args['restOnExecutionStartFunction'] = function (&$arguments) use ($requestHash) {
-            $this->updateRetryHeaders(
-                $arguments,
-                $requestHash
+        // Adding callback logic to update headers while retrying
+        $args['restRetryListener'] = function (
+            \Exception $e,
+            $retryAttempt,
+            $arguments
+        ) use (
+            $invocationIdHeaderValue,
+            $attempCountKey
+) {
+            $arguments[0] = $this->addRetryListenerCallback(
+                $retryAttempt,
+                $arguments[0],
+                $invocationIdHeaderValue,
+                $attempCountKey
             );
+            return $arguments;
         };
         return $args;
     }
 
-    /**
-     * Updates the api client identification header value with UUID
-     * and retry count
-     *
-     * @param array &$arguments The arguments array(passed by reference) used by
-     * execute method of ExponentialBackoff object.
-     * @param string $requestHash A UUID4 string value that represents a request and
-     * it's retries.
-     * @param int $currentAttempt The original attempt is of a value 1, and retries are
-     * 2, 3 and so on.
-     * @return void
-     */
-    private function updateRetryHeaders(
-        &$arguments,
-        $requestHash,
-        $currentAttempt = 1
+    private function addRetryListenerCallback(
```

Comment:
No docbloc.
I think this method is not named correctly? I don't see a retry listener being attached.

If this method is just a proxy to calling appendOrModifyHeaders, then we probably don't need it?

I don't think we need to pass `$attemptCountKey` everywhere. Let's just declare a constant and use it in the place where we need it.